src/iterative_ensemble_smoother/esmda.py

```diff
@@ -0,0 +1,256 @@
+"""
+Ensemble Smoother with Multiple Data Assimilation (ES-MDA)
+----------------------------------------------------------
+
+Imlementation of the 2013 paper "Ensemble smoother with multiple data assimilation"
+
+References
+----------
+
+Emerick, A.A., Reynolds, A.C. History matching time-lapse seismic data using
+the ensemble Kalman filter with multiple data assimilations.
+Comput Geosci 16, 639–659 (2012). https://doi.org/10.1007/s10596-012-9275-5
+
+Alexandre A. Emerick, Albert C. Reynolds.
+Ensemble smoother with multiple data assimilation.
+Computers & Geosciences, Volume 55, 2013, Pages 3-15, ISSN 0098-3004,
+https://doi.org/10.1016/j.cageo.2012.03.011.
+
+https://gitlab.com/antoinecollet5/pyesmda
+
+https://helper.ipam.ucla.edu/publications/oilws3/oilws3_14147.pdf
+
+"""
+
+import numbers
+from typing import Optional, Union
+
+import numpy as np
+import numpy.typing as npt
+import scipy as sp  # type: ignore
+
+from iterative_ensemble_smoother.esmda_inversion import (
+    inversion_exact_cholesky,
+    inversion_subspace,
+    normalize_alpha,
+)
+
+
+class ESMDA:
+    # Available inversion methods. The inversion methods all compute
+    # C_MD @ (C_DD + alpha * C_D)^(-1)  @ (D - Y)
+    _inversion_methods = {
+        "exact": inversion_exact_cholesky,
+        "subspace": inversion_subspace,
+    }
+
+    def __init__(
+        self,
+        C_D: npt.NDArray[np.double],
+        observations: npt.NDArray[np.double],
+        alpha: Union[int, npt.NDArray[np.double]] = 5,
+        seed: Union[np.random._generator.Generator, int, None] = None,
+        inversion: str = "exact",
+    ) -> None:
+        """Initialize Ensemble Smoother with Multiple Data Assimilation (ES-MDA).
+
+        The implementation follows the 2013 paper by Emerick et al.
+
+        Parameters
+        ----------
+        C_D : np.ndarray
+            Covariance matrix of outputs of shape (num_outputs, num_outputs).
+            If a 1D array is passed, it represents a diagonal covariance matrix.
+        observations : np.ndarray
+            1D array of shape (num_inputs,) representing real-world observations.
+        alpha : int or 1D np.ndarray, optional
+            If an integer `alpha` is given, an array with length `alpha` and
+            elements `alpha` is constructed. If an 1D array is given, it is
+            normalized so sum_i 1/alpha_i = 1 and used. The default is 5, which
+            corresponds to np.array([5, 5, 5, 5, 5]).
+        seed : integer, optional
```

Comment:
Type does not correspond to type-hint which is why I in general don't like documenting type-hints.